Move Parse function to JSON; add NDT coordination logic.

collina opened the pull request:
<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/m-lab/ndt-javascript/6)

<!-- Reviewable:end -->

Commit:
Move parser to JSON.

Commit:
Move Parse function to JSON; add NDT coordination logic.

mtlynch commented:
Review status: 0 of 2 files reviewed at latest revision, 21 unresolved discussions.

---

_[src/ndt-client.js, line 33 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9SZ66WQbHQ2t2oAdAX:-K9SZ66WQbHQ2t2oAdAY:-949171984) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L33)):_
Can we document this constant? It's unexpected that the client would fix a server version anywhere. Is this the min required version, max, or does the server have to be exactly this version?

---

_[src/ndt-client.js, line 39 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9SYKB8qM-5ItEImbJN:-K9SYKB8qM-5ItEImbJO:-690803045) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L39)):_
Can we create a file that just defines the protocol constants so that we don't have to use magic numbers? By "protocol constants", I mean the values we have to choose because they're defined in:

https://github.com/ndt-project/ndt/blob/master/src/protocol.h
or
https://github.com/ndt-project/ndt/blob/master/src/ndtptestconstants.h

---

_[src/ndt-client.js, line 253 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9ScJxtY3dyqgt1_u6v:-K9ScJxu5XteGfgkpG1j:231785402) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L253)):_
Recommend not saying specifically what tests because this will change if we redefine `NDT_DESIRED_TESTS`.

---

_[src/ndt-client.js, line 256 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9Scka8ze48qKNWtlpt:-K9Scka8ze48qKNWtlpu:704607651) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L256)):_
Can this be a defined constant?

Ditto for pretty much every magic string in this function.

---

_[src/ndt-client.js, line 259 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9Sd9tJhMQNREMNcWiJ:-K9Sd9tJhMQNREMNcWiK:740212502) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L259)):_
This is a pretty long, complex function. Can we break it up?

---

_[src/ndt-client.js, line 265 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9SdLAT-p-x7Yn7Adfd:-K9SdLAT-p-x7Yn7Adfe:-830858888) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L265)):_
`parseNDTMessage` never checks that the message is a valid type, so it seems like this logging call could throw an exception, which seems undesirable. Can we validate the type in `parseNDTMessage`?

---

_[src/ndt-client.js, line 271 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9Sf_8VNWpbRp07upUa:-K9Sf_8VNWpbRp07upUb:-1247605656) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L271)):_
What happens if it doesn't return `'DONE'`? It doesn't seem like it has any way of continuing.

---

_[src/ndt-client.js, line 272 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9SepKpx5ik6-w9kTDk:-K9SepKqvop7VfF34T0B:-140867936) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L272)):_
Having to define and undefine a variable feels kind of ugly. Can we just have the variable be a bool like 

``` javascript
var testInProgress = false;
...
if (!testInProgress && remainingTests.length > 0) {
  testInProgress = true;
  if (remainingTests.pop()(parsedMessage) === 'DONE') ...
```

---

_[src/ndt-client.js, line 278 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9SgBpWmALx0HiQVf3W:-K9SgBpWmALx0HiQVf3X:1043322903) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L278)):_
The implementation doesn't seem to agree with this description. If the first message we get back from the server is `MSG_LOGIN`, we seem to forget that we ever wanted to see `SRV_QUEUE`.

---

_[src/ndt-client.js, line 279 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9Sfsuae7eRYSSSvb5K:-K9SfsubYY9q4MEVus96:-61523253) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L279)):_
Can we clarify this comment? It's not clear to me what `SRV_QUEUE("0")` means.

---

_[src/ndt-client.js, line 283 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9SgtQ5wERGP6Ao0zjM:-K9SgtQ5wERGP6Ao0zjN:1445984451) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L283)):_
How frequently do we expect to call `makeNDTMessage`? Consider wrapper functions to simplify the calling semantics, so we can just do `makeNDTWaitingMessage`.

---

_[src/ndt-client.js, line 289 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9ShMMwIqbgd1F0Uvnu:-K9ShMMwIqbgd1F0Uvnv:1003200515) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L289)):_
Where are we implementing the wait? The log message seems more appropriate there.

---

_[src/ndt-client.js, line 292 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9ShiDzOHrlSyYLANXE:-K9ShiDzOHrlSyYLANXF:1235793489) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L292)):_
Why are we not handling the "0" or "9988" cases explicitly?
https://github.com/ndt-project/ndt/wiki/NDTProtocol#queuing-clients

---

_[src/ndt-client.js, line 294 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9Sjpjf3OEO8_Rf9VCB:-K9Sjpjf3OEO8_Rf9VCC:1169203249) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L294)):_
I'm assuming that the implication is that any server capable of talking to us is compatible so we don't need to check the version number? We should document this though.

---

_[src/ndt-client.js, line 314 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9SjEy0IUpd-B4PKKV5:-K9SjEy0IUpd-B4PKKV6:-325052824) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L314)):_
It doesn't seem like we're implementing this part of the spec:

> The Client MUST drop the connection when it receives unknown test id, because such situation means an error on the Server's part.

https://github.com/ndt-project/ndt/wiki/NDTProtocol#negotiating-test-suite

---

_[src/ndt-client.js, line 315 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9SojfqsR3XnBktNCRe:-K9SojfqsR3XnBktNCRf:-1570785696) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L315)):_
I'm not following how this works. How do we go immediately from finding out which tests we're going to run into getting the results of the tests. Feels like there should be a `RUN_TESTS` state in there.

---

_[src/ndt-client.js, line 318 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9SofgDfArebGRvD96N:-K9SofgEzvRCDiOa_1LU:1319562281) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L318)):_
Is this a TODO?

---

_[src/ndt-client.js, line 337 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9SpCYCD1uiyfsZK4pF:-K9SpCYCD1uiyfsZK4pG:-1898843780) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L337)):_
Do these need TODOs or is this the whole implementation?

---

_[test/ndtSpec.js, line 29 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9SqO4jZ3LLj9aEPW1W:-K9SqO4kSZQxkwMgvfge:1543061714) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/test/ndtSpec.js#L29)):_
Can we replace magic number (ditto for other tests)

---

_[test/ndtSpec.js, line 111 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9SnHYojyAKVLz9EOYY:-K9SnHYojyAKVLz9EOYZ:-1404958701) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/test/ndtSpec.js#L111)):_
Is this testing against a production server? I _really_ don't think we should do that.

---

_[test/ndtSpec.js, line 117 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9SnTZgvmVeyaMrunnd:-K9SnTZgvmVeyaMrunne:1435299252) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/test/ndtSpec.js#L117)):_
I think we need WAY more test coverage. Just some example scenarios I think we need to cover:
- Server sends malformed version string
- Server sends `MSG_LOGIN` without sending `SRV_QUEUE`
- Server sends message with illegal type
- Server sends an unexpected message in response to login
- Server sends an invalid test ID
- Server does not honor the tests the client requested

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6)_

<!-- Sent from Reviewable.io -->

collina commented:
Review status: 0 of 2 files reviewed at latest revision, 21 unresolved discussions.

---

_[src/ndt-client.js, line 39 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9SYKB8qM-5ItEImbJN:-KADLBt7M2-_5yy5xTGK:-1176835252) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L39)):_
I would prefer not, creating an additional file that has to be compiled in through grunt or other build mechanisms will add unnecessary complexity to distribution and use, especially for only a few constants. 

---

_[src/ndt-client.js, line 265 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9SdLAT-p-x7Yn7Adfd:-KADUsSP6azkCRoF1gmh:2106604441) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L265)):_
Message type is sequential and numerical, so we rely on it not being out of bounds. I've added what I think is as close of a check as I can anticipate. 

---

_[src/ndt-client.js, line 271 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9Sf_8VNWpbRp07upUa:-KADZhBpOSpuRn_Z3wPg:-1787613324) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L271)):_
It should fail in that case, correct? Added an else statement to fail.

---

_[src/ndt-client.js, line 283 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9SgtQ5wERGP6Ao0zjM:-KADbJ-4uwNCnO95-PAD:-221405070) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L283)):_
I'm not sure though that we need to create a wrapper for a message that we call once, I believe that `makeNDTMessage` will be called multiple times with different states.

---

_[src/ndt-client.js, line 289 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9ShMMwIqbgd1F0Uvnu:-KADbn8PJN0TyfM1evEA:692351917) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L289)):_
Do we need to wait on the client side though? I think the client will wait for the NDT server will emit another message, and that the client doesn't need to sleep?

---

_[src/ndt-client.js, line 315 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9SojfqsR3XnBktNCRe:-KADdO1p4zNMbVJJdy5z:-121981946) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L315)):_
By this point, `remainingTests` should be populated and ndtState is WAIT_FOR_MSG_RESULTS, however, it's interdicted by that first if that interates across `remainingTests` until depletion.

---

_[src/ndt-client.js, line 318 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9SofgDfArebGRvD96N:-KADOhXsFcUSO-g8GXqo:525712321) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L318)):_
Yes, wanders a bit down another path so it's a placeholder.

---

_[test/ndtSpec.js, line 29 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9SqO4jZ3LLj9aEPW1W:-KAD_P4o8Lj1Qccv9L5G:1668803948) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/test/ndtSpec.js#L29)):_
Here though the magic number doesn't have any more impact than the 'OK' string, we aren't checking the state in this function. It seems like this adds unnecessary complexity. 

---

_[test/ndtSpec.js, line 117 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9SnTZgvmVeyaMrunnd:-KADcC1Sp7eG3UXioqVy:869711772) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/test/ndtSpec.js#L117)):_
Ack. I need to separate this into another PR, since the WebSocket unit test writing is going to be an expansive mess.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6)_

<!-- Sent from Reviewable.io -->

Commit:
Address PR feedback on commenting and constants.

mtlynch commented:
Review status: 0 of 2 files reviewed at latest revision, 21 unresolved discussions.

---

_[src/ndt-client.js, line 39 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9SYKB8qM-5ItEImbJN:-KAG57w_znbVSF9LtgfE:319770748) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L39)):_
I'd like a separate file because I think it's cleaner, but I don't see it as critical. Can you talk more about the cost of adding source files? Is that something that's just in general maintenance-expensive in node or are we using particular libraries/build tools that make it expensive?

The important part is getting rid of the magic numbers and making it clear where they come from. We're currently using almost every `#define` from https://github.com/ndt-project/ndt/blob/master/src/protocol.h

By my count, it's 24 constants (12 message types, 5 server queue message types, and 7 test types), which seems like a pretty significant number to me.

Can we clearly demarcate the constants that we're defining because we have to match something in NDT? I think the current organization is a bit confusing because we have `messageType` and we assign it a bunch of numbers to match `procotol.h`, but then right after that, we have `testState` and we're assigning numbers that we choose ourselves, so I think this can be confusing for the reader.

---

_[src/ndt-client.js, line 259 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9Sd9tJhMQNREMNcWiJ:-KAG9o2lV8aFd_hf_3Qa:971039376) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L259)):_
ping on this one. I'm not clear on what the resolution was.

---

_[src/ndt-client.js, line 289 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9ShMMwIqbgd1F0Uvnu:-KAGG_Xo0LkHMyEKxK3a:1714413084) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L289)):_
Oh, then can we clarify the language? The current text causes me to think that we're going to explicitly pause ourselves.

Suggestion: `Received wait notice: server estimates %d minutes before test will begin`

---

_[src/ndt-client.js, line 68 [r2]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-KAG7x3HoqmM4Svc8Z8V:-KAG7x3HoqmM4Svc8Z8W:1050001577) ([raw file](https://github.com/m-lab/ndt-javascript/blob/f9db5280b7a08a6ce5e9f8c6fa456b9bb04fccb9/src/ndt-client.js#L68)):_
This is an enum, right? Can we give it enum type and naming conventions?
https://google.github.io/styleguide/javascriptguide.xml?showone=JavaScript_Types#JavaScript_Types

---

_[src/ndt-client.js, line 72 [r2]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-KAGAWvE-D0vu4c2ZzsQ:-KAGAWvFOFXXRjgSZFd_:-653837926) ([raw file](https://github.com/m-lab/ndt-javascript/blob/f9db5280b7a08a6ce5e9f8c6fa456b9bb04fccb9/src/ndt-client.js#L72)):_
It seems like there's an implicit assumption here that this list be in exactly the same order as `messageType`. Can we get rid of this assumption by making this a dictionary like this?

```
messageName: {
  messageType.COMM_FAILURE: 'COMM_FAILURE',
  ...
```

---

_[src/ndt-client.js, line 218 [r2]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-KAG97rQcruWrJwiIe8x:-KAG97rQcruWrJwiIe8y:-365146688) ([raw file](https://github.com/m-lab/ndt-javascript/blob/f9db5280b7a08a6ce5e9f8c6fa456b9bb04fccb9/src/ndt-client.js#L218)):_
I don't know if this is just my JavaScript idiocy showing but I really don't understand this line. Isn't `NDTMessage.type` a `Uint8` and `MESSAGE_NAME` is a list of strings?

---

_[src/ndt-client.js, line 291 [r2]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-KAGGDZH5OSdYOrLl6za:-KAGGDZH5OSdYOrLl6zb:1660794408) ([raw file](https://github.com/m-lab/ndt-javascript/blob/f9db5280b7a08a6ce5e9f8c6fa456b9bb04fccb9/src/ndt-client.js#L291)):_
Can we re-word a bit to clarify who is sending and who is receiving?

---

_[src/ndt-client.js, line 292 [r2]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-KAGBgZYEDOyA64DJJsI:-KAGBgZYEDOyA64DJJsJ:1918261175) ([raw file](https://github.com/m-lab/ndt-javascript/blob/f9db5280b7a08a6ce5e9f8c6fa456b9bb04fccb9/src/ndt-client.js#L292)):_
Can we call this `SRV_QUEUE_TEST_STARTS_NOW` (then the English explanation becomes unnecessary)
https://github.com/ndt-project/ndt/wiki/NDTProtocol#queuing-clients

---

_[src/ndt-client.js, line 295 [r2]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-KAGBpRYK4ADuE1fBvv3:-KAGBpRYK4ADuE1fBvv4:1958467974) ([raw file](https://github.com/m-lab/ndt-javascript/blob/f9db5280b7a08a6ce5e9f8c6fa456b9bb04fccb9/src/ndt-client.js#L295)):_
I'd like to get rid of these magic strings as well and declare them alongside the other constants from `protocol.h`

---

_[src/ndt-client.js, line 302 [r2]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-KAGGOufenjh09sH8jdz:-KAGGOufenjh09sH8je-:-593209838) ([raw file](https://github.com/m-lab/ndt-javascript/blob/f9db5280b7a08a6ce5e9f8c6fa456b9bb04fccb9/src/ndt-client.js#L302)):_
I think if we switch to symbolic names for the messages, these comments become unnecessary

---

_[src/ndt-client.js, line 314 [r2]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-KAGHFzgd2aoslvZ_XfO:-KAGHFzhnlEfBN4ZRDmo:-1821631768) ([raw file](https://github.com/m-lab/ndt-javascript/blob/f9db5280b7a08a6ce5e9f8c6fa456b9bb04fccb9/src/ndt-client.js#L314)):_
nitpick: JavaScript

---

_[src/ndt-client.js, line 329 [r2]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-KAGJ3QUQC7nOlktC1i9:-KAGJ3QVtKMo2nH8Qzj9:601430560) ([raw file](https://github.com/m-lab/ndt-javascript/blob/f9db5280b7a08a6ce5e9f8c6fa456b9bb04fccb9/src/ndt-client.js#L329)):_
Can we get rid of these magic strings as well?

---

_[src/ndt-client.js, line 333 [r2]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-KAGJPzg0znKAf9-kgqq:-KAGJPzg0znKAf9-kgqr:100034983) ([raw file](https://github.com/m-lab/ndt-javascript/blob/f9db5280b7a08a6ce5e9f8c6fa456b9bb04fccb9/src/ndt-client.js#L333)):_
Can we make a dictionary of test IDs to tests and reduce this to:

```
parsedTestIds.forEach(function(testId) {
  if (!(testId in testIdToTest)) {
    throw Error('ReceivedUnsupportedTest');
  }
  remainingTests.push(testIdToTest[testId](ndtControlSocket));
}
```

---

_[src/ndt-client.js, line 370 [r2]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-KAGM94C7GhEExxXpblS:-KAGM94DSXiJRihQK1t3:1925825635) ([raw file](https://github.com/m-lab/ndt-javascript/blob/f9db5280b7a08a6ce5e9f8c6fa456b9bb04fccb9/src/ndt-client.js#L370)):_
Minor: Can we use the syntax `TODO`? I'm used to seeing that and I can immediately recognize what it means, but I've never seen `[To Complete]`

---

_[test/ndtSpec.js, line 29 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9SqO4jZ3LLj9aEPW1W:-KAGLdoqu-EXaMvq7lem:-1769710972) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/test/ndtSpec.js#L29)):_
Doesn't `parseNDTMessage` require the message type to be a known type now? The number's not entirely meaningless because if we made this `buildMessageBuffer(100, 'OK')` this should fail if I'm understanding correctly. 

Suggest using a real message type (from our defined constants).

---

_[test/ndtSpec.js, line 117 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9SnTZgvmVeyaMrunnd:-KAGMf8hTuzTEh-wR2LJ:1967967563) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/test/ndtSpec.js#L117)):_
I really want to avoid splitting up production code PRs from unit test PRs. I think the options we have are:
1. Split the functionality of this PR into smaller chunks that we can unit test separately without bloating the PR
2. Put this PR on hold and do a separate PR that implements the general mechanisms we need for implementing server mocking, submit that, then rebase this PR onto that one and complete the unit tests
3. Just make a really large PR
4. Defer unit tests

(4) is my least favorite. Can we rebase this PR onto the coveralls PR? If the code coverage results from coveralls are going to keep us honest with respect to which test cases we're missing, I'm more open to (4) but it's still my least desirable.

---

_[test/ndtSpec.js, line 111 [r2]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-KAGKsADcMO-JXhh_6nm:-KAGKe6WHHQDoS8kIBQw:320278841) ([raw file](https://github.com/m-lab/ndt-javascript/blob/f9db5280b7a08a6ce5e9f8c6fa456b9bb04fccb9/test/ndtSpec.js#L111)):_
This is probably just my lack of familiarity with karma and maybe doesn't require documentation in the code itself, but could you help me understand how the code knows to not really hit this URL under test? I can't see anything that seems to mock this out.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6)_

<!-- Sent from Reviewable.io -->

collina commented:
Review status: 0 of 2 files reviewed at latest revision, 19 unresolved discussions.

---

_[src/ndt-client.js, line 39 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9SYKB8qM-5ItEImbJN:-KALT6wTGlZWMEPbH42v:1014125314) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L39)):_
I am happy for ridding us of magic numbers scattered about, but multiple files requires either some compilation routine, requirejs or inclusion of further scripts within the embedding site.  Which are equal in added work. If we determine this is necessary, then we should discuss it in the bug tracker rather than this PR, since it was code that was already reviewed. I would rather not conflate decisions approved in prior rounds with the specific changes made under this PR. 

---

_[src/ndt-client.js, line 259 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9Sd9tJhMQNREMNcWiJ:-KALV2dgO9yLxaCGrbaa:-1557948373) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L259)):_
As I mentioned in the CR email reply, this was next after the others are resolved because it's a move of code that has been more heavily commented. 

---

_[src/ndt-client.js, line 72 [r2]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-KAGAWvE-D0vu4c2ZzsQ:-KALWPVCc4LXe8ACpYCw:787326243) ([raw file](https://github.com/m-lab/ndt-javascript/blob/f9db5280b7a08a6ce5e9f8c6fa456b9bb04fccb9/src/ndt-client.js#L72)):_
It's not exactly implicit; consistency is checked within the unit tests for NDTjs. Same notion as well for #43

---

_[src/ndt-client.js, line 218 [r2]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-KAG97rQcruWrJwiIe8x:-KAL_-ZQmae1sswdii7p:2096983479) ([raw file](https://github.com/m-lab/ndt-javascript/blob/f9db5280b7a08a6ce5e9f8c6fa456b9bb04fccb9/src/ndt-client.js#L218)):_
I do not believe it is `Uint8`, and testing didn't seem to indicate so but I will explicitly set the type to ensure that it is a Number.

---

_[src/ndt-client.js, line 333 [r2]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-KAGJPzg0znKAf9-kgqq:-KALz6R1EsE9D0r07EQH:-1510874628) ([raw file](https://github.com/m-lab/ndt-javascript/blob/f9db5280b7a08a6ce5e9f8c6fa456b9bb04fccb9/src/ndt-client.js#L333)):_
The different subtests call out to their functions in different ways, so I would like to keep these different, but I have simplified the client code, and used unit tests to verify alignment of variables.

---

_[src/ndt-client.js, line 370 [r2]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-KAGM94C7GhEExxXpblS:-KALUjSGYNJwZpjYZ50b:-1605377872) ([raw file](https://github.com/m-lab/ndt-javascript/blob/f9db5280b7a08a6ce5e9f8c6fa456b9bb04fccb9/src/ndt-client.js#L370)):_
These are placeholder comments that are necessary for a final script, and the next series of revisions for the code review process after unit tests. I'm not sure there is a standardization, so I think I think it is fair to leave it to discretion. 

---

_[test/ndtSpec.js, line 117 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9SnTZgvmVeyaMrunnd:-KALfJ5jXo6hJkb1FKmp:-1401289197) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/test/ndtSpec.js#L117)):_
Understood, but I anticipate that the fake WebSocket object that will emulate an NDT server is going to be a meaningful amount of work, and I would rather not have a large PR. I do not believe 1 is possible for that matter, I believe this quickly goes to a state of bloat. 

---

_[test/ndtSpec.js, line 111 [r2]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-KAGKsADcMO-JXhh_6nm:-KALeXPgSYtflb6_n9ts:-956086718) ([raw file](https://github.com/m-lab/ndt-javascript/blob/f9db5280b7a08a6ce5e9f8c6fa456b9bb04fccb9/test/ndtSpec.js#L111)):_
It does but it is only a value to pass while we instantiate NDTjs, which does not connect until the test is started -- this does not occur within the unit tests currently, Where Websockets is actually triggered, it will be done using a fake WebSockets object that acts as if the server would, which is going to be a heavy lift and the reason why there is not substantial unit testing at the moment.  

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6)_

<!-- Sent from Reviewable.io -->

Commit:
Remove magic numbers, simplify test thread, reply to comments.

mtlynch commented:
Review status: 0 of 2 files reviewed at latest revision, 14 unresolved discussions.

---

_[src/ndt-client.js, line 39 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9SYKB8qM-5ItEImbJN:-KAMEIV54dYVHkQOmDmM:-1928896013) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/src/ndt-client.js#L39)):_
Okay separate file sounds like more trouble than it's worth.

Can we split out the constants that are defined to match values in https://github.com/ndt-project/ndt/wiki/NDTProtocol to a separate namespace where we document that that's where the values are coming from?

I think the current organization can lead to confusion because `TestState` is not based on any values defined by the protocol itself (I don't think), but it's right between two constants that are.

I'm fine with filing a bug on this to handle this outside this PR if you feel this is out of scope.

---

_[src/ndt-client.js, line 72 [r2]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-KAGAWvE-D0vu4c2ZzsQ:-KAMCb3lBI_19dYH-rQ-:856463378) ([raw file](https://github.com/m-lab/ndt-javascript/blob/f9db5280b7a08a6ce5e9f8c6fa456b9bb04fccb9/src/ndt-client.js#L72)):_
The unit tests check the length, but don't verify order, so if we swapped the order of two elements, we'd print incorrect names. Also, this is another place where we break if someone adds a non-consecutive message type to NDT.

I can file this as a bug if you'd prefer to make the change outside this PR.

---

_[src/ndt-client.js, line 218 [r2]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-KAG97rQcruWrJwiIe8x:-KAMHUR_xlBu267KWy_-:1478131130) ([raw file](https://github.com/m-lab/ndt-javascript/blob/f9db5280b7a08a6ce5e9f8c6fa456b9bb04fccb9/src/ndt-client.js#L218)):_
I think I was unclear in my earlier note. The original code was this:

`this.constants.MESSAGE_NAME.indexOf(NDTMessage.type) != -1`

So just as an example, if the message type was 5, the interpreter would basically see this:

`['COMM_FAILURE', 'SRV_QUEUE', ...].indexOf(5)`

Shouldn't that always fail? Or am I misunderstanding?

The current code looks correct, but if I haven't made a mistake above, I'm worried about our unit tests now. Because if the unit tests for this still passed on the earlier version of this code, I'd like to check what's going on with our unit tests because it seems like we might be getting false positives.

---

_[src/ndt-client.js, line 58 [r3]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-KAMNbhQ8dHbezCrz4CO:-KAMNbhQ8dHbezCrz4CP:1562442930) ([raw file](https://github.com/m-lab/ndt-javascript/blob/04e502b69349ff0bd11f938c713e98648d134634/src/ndt-client.js#L58)):_
Can these be numbers so we match the defined NDT protocol? It seems like other parts of the client should be responsible for serializing/deserializing these between numbers to strings.

Also since there are only 6 total defined test types, suggest defining all the values in the protocol for completeness and calling this `TestTypes`. `SUPPORTED_TEST_IDS` will still make it clear which ones the client supports.

---

_[src/ndt-client.js, line 60 [r3]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-KAMNIoNiWVUtPN7uUaW:-KAMNIoO4y1u2AnV_IXp:1102373214) ([raw file](https://github.com/m-lab/ndt-javascript/blob/04e502b69349ff0bd11f938c713e98648d134634/src/ndt-client.js#L60)):_
Can we use the symbolic constants so we're not duplicating strings?

Also, I at first thought `SUPPORTED_TEST_IDS` and `NDT_DESIRED_TESTS` were redundant, but then I realized it's good that we make a distinction between the tests we support and the tests we want to run. Today they're the same, but in the future, we probably want to allow users to run a subset of tests (when the server supports it). Can we document `SUPPORTED_TEST_IDS` to make the distinction clear?

---

_[src/ndt-client.js, line 236 [r3]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-KAMJrFnzRGWMq9Qvlmy:-KAMJrFoTuyuGrP8uVlC:-1656029666) ([raw file](https://github.com/m-lab/ndt-javascript/blob/04e502b69349ff0bd11f938c713e98648d134634/src/ndt-client.js#L236)):_
`MESSAGE_NAMES` is a variable we derived from `MessageType` but `MessageType` is the authoritative definition of these, so I'd like to base the check on `MessageType`.

This check would also break if the NDT protocol defines a message that has a non-consecutive ID (a not outlandish possibility if someone decides to define a new class of messages starting at like `100`).

Given that we do the same type of check on both `MessageType` and  `SupportedTests`, what do you think of implementing a `hasOwnValue` as described here?

http://jamesallardice.com/how-to-check-if-a-javascript-object-has-a-certain-value/

---

_[src/ndt-client.js, line 356 [r3]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-KAMPNuNP3FNCYVuLFGY:-KAMPNuO0h0nYrkYt_9U:-16919603) ([raw file](https://github.com/m-lab/ndt-javascript/blob/04e502b69349ff0bd11f938c713e98648d134634/src/ndt-client.js#L356)):_
The NDT spec is kind of unclear on what the client does if the server specifies tests that the client didn't offer.

>  This list of tests's ids MUST NOT contain any tests that were not requested by the Client in the first MSG_LOGIN message. 
> ...
> The Client MUST drop the connection when it receives unknown test id, because such situation means an error on the Server's part.

https://github.com/ndt-project/ndt/wiki/NDTProtocol#negotiating-test-suite

Right now it's not an issue, but in the future we may want to support letting the client run a subset of tests it's compatible with. So if the client requests just `TEST_S2C` and the server comes back with `TEST_C2S`, that shouldn't trick the client into running `TEST_C2S`.

This is a longwinded way of saying I think we should check this by &'ing against `NDT_DESIRED_TESTS`.

---

_[test/ndtSpec.js, line 117 [r1]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-K9SnTZgvmVeyaMrunnd:-KAMUOiNcAG1pL1UnYHq:1154907976) ([raw file](https://github.com/m-lab/ndt-javascript/blob/89a13b48808fe84f5ac980ea4982db114d3bbb68/test/ndtSpec.js#L117)):_
Per offline discussion, sounds like we're going to do (3).

---

_[test/ndtSpec.js, line 51 [r3]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-KAMTXIaFawz7MtYvZWH:-KAMTXIaFawz7MtYvZWI:1952811418) ([raw file](https://github.com/m-lab/ndt-javascript/blob/04e502b69349ff0bd11f938c713e98648d134634/test/ndtSpec.js#L51)):_
Same note about not relying on the assumption that message types will always be consecutive numbers. I think just choosing some arbitrary number like `999` is good.

---

_[test/ndtSpec.js, line 96 [r3]](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6#-KAMUEtfggzauDi7OJvc:-KAMUEtfggzauDi7OJvd:26080070) ([raw file](https://github.com/m-lab/ndt-javascript/blob/04e502b69349ff0bd11f938c713e98648d134634/test/ndtSpec.js#L96)):_
I don't think we'll need this test at all if we switch to defining `SUPPORTED_TEST_IDS` with constants.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/m-lab/ndt-javascript/6)_

<!-- Sent from Reviewable.io -->
